### ISSUE TITLE
How to initialize and apply an ensemble of models with same input?

### ISSUE THREAD

**Issue by jjyyxx** [deepmind/dm-haiku#533]
In a haiku module, I would like to create N networks with the same structure but different parameters, then apply the same input to the N networks to get a stacked output. My current solution is using a for-loop (liist comprehension, see below) to init and apply models, but I wonder if functional (vmap over parameters, possibly) transform could apply here.

```python
# In class A(hk.Module)
# __init__
self.models = [hk.Seqential(..., name=f“model_{i}”) for i in range(N)]
## Something like hk.lift. hk.transform, hk.vmap, jax.vmap?

# __call__(x)
y = [model(x) for model in models]  # x in shape of (B, C)
y = jnp.stack(y, axis=-2)  # y in shape of (B, N, D)
```

I was aware of [hk.lift documentation](https://dm-haiku.readthedocs.io/en/latest/notebooks/transforms.html#Using-hk.lift), but the documentation, especially snippets in the summary part, does not seem to be fully working, making me more confused.

**Comment by nlsfnr** [comment]
Yes. You can also use `jax.pmap` instead of `vmap` to run one model per device.

```python
import haiku as hk
import jax.numpy as jnp
import jax


def model(x: jax.Array) -> jax.Array:
    weights = hk.get_parameter('w',
                               (x.shape[-1], 10),
                               init=hk.initializers.RandomNormal())
    return x @ weights


def main() -> None:

    USE_PMAP = False
    N_MODELS = jax.device_count() if USE_PMAP else 10

    rng = jax.random.PRNGKey(0)
    model_hk = hk.transform(model)
    model_hk = hk.without_apply_rng(model_hk)
    inputs = jnp.ones((N_MODELS, 1))
    rngs = jax.random.split(rng, N_MODELS)
    params = (jax.pmap if USE_PMAP else jax.vmap)(model_hk.init)(rngs, inputs)
    assert all(param.shape[0] == N_MODELS
               for param in jax.tree_util.tree_leaves(params))
    outputs = jax.vmap(model_hk.apply)(params, inputs)
    assert all(output.shape[0] == N_MODELS
               for output in jax.tree_util.tree_leaves(outputs))
    breakpoint()


if __name__ == '__main__':
    main()
```

**Comment by jjyyxx** [comment]
@nlsfnr Thanks for your suggestion. The `vmap` outside `transform` way is what I am familiar with. However, the problem is that I want to achieve this within a `hk.Module`, to be assembled into a larger network and transform the top-level network. The snippet in my issue described the ideal interface, but I wish to use `vmap` instead of for-loop.

Also, that's why I'm looking into the [hk.lift documentation](https://dm-haiku.readthedocs.io/en/latest/notebooks/transforms.html#Using-hk.lift), with no success.

**Comment by tomhennigan** [comment]
Hi @jjyyxx, apologies for the delay getting back to you.

I agree with @nlsfnr that JAX transforms outside of Haiku is probably the safest route to implement this, but I understand that there are cases where this would require significant refactoring and so doing the transformation inside `hk.transform` is preferred.

This is a case where `hk.lift` is the right tool. Here is an example following the structure you proposed: https://colab.research.google.com/gist/tomhennigan/828ad6ade44713c487afd88c042a27dc/example-of-ensembling-models-in-haiku.ipynb#scrollTo=UALB9Ug2VLNY

The key part of the notebook is the `create_ensemble` function which drives `lift` for you:

```python
def create_ensemble(model, ensemble_size):
  init_rng = hk.next_rng_keys(ensemble_size) if hk.running_init() else None
  model = hk.transform(model)
  # in_axes: rng is mapped, data is not.
  init_model = hk.lift(jax.vmap(model.init, in_axes=(0, None)),
                       name="ensemble")
  def ensemble(x):
    params = init_model(init_rng, x)
    # in_axes: params are mapped, rng/data are not.
    return jax.vmap(model.apply, in_axes=(0, None, None))(params, None, x)
  return ensemble
```

You can then use this to ensemble whatever network you prefer:

```python
def model(x):
  return hk.nets.MLP([300, 100, 10])(x)

def f(x):
  model = create_ensemble(model, ensemble_size=4)
  return model(x)

f = hk.transform(f)
```

Of course this can be put inside another Haiku module:

```python
def model(x):
  return hk.nets.MLP([300, 100, 10])(x)

class A(hk.Module):
  def __init__(self):
    super().__init__()
    self.models = create_ensemble(model, ensemble_size=4)

  def __call__(self, x):
    return self.models(x)

def f(x):
  return A()(x)

f = hk.transform(f)
```

> I was aware of [hk.lift documentation](https://dm-haiku.readthedocs.io/en/latest/notebooks/transforms.html#Using-hk.lift), but the documentation, especially snippets in the summary part, does not seem to be fully working, making me more confused.

I tried running the snippets in the docs and they are working for me in Colab, please let me know what is not working for you and I will take a look.

**Comment by jjyyxx** [comment]
Thanks for your help @tomhennigan ! Your solution works perfectly. I think it's a good example to demonstrate vmapping over parameters and could be added to the `hk.lift` documentation.

My final snippet is
```python
from typing import Optional

import jax
import haiku as hk


def create_ensemble(model: hk.Module, ensemble_size: int, name: Optional[str] = None):
    init_rng = hk.next_rng_keys(ensemble_size) if hk.running_init() else None
    transformed = hk.transform(model)
    init_model = hk.lift(jax.vmap(transformed.init, in_axes=(0, None)), name=name)
    def ensemble(x):
        params = init_model(init_rng, x)
        return jax.vmap(transformed.apply, in_axes=(0, None, None), out_axes=-2)(params, None, x)
    return ensemble
```
I used `out_axes=-2` to put the ensemble axis after the batch axes (to fit the shape of other parts of my code). The substitution also leads to a 50% performance boost (of the module itself).

> I was aware of [hk.lift documentation](https://dm-haiku.readthedocs.io/en/latest/notebooks/transforms.html#Using-hk.lift), but the documentation, especially snippets in the summary part, does not seem to be fully working, making me more confused.

Sorry for this false statement. The document is correct. I found myself misunderstanding the document of `hk.lift`.
